#include "mainwindow.h"
#include "ui_mainwindow.h"

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    // Á™óÂè£„ÄÅËßÜÂå∫ËÆæÁΩÆ
    scene = new MainScene();
    view = new MainView(scene);
    setCentralWidget(view);         // Â∞Ü view ËÆæÁΩÆ‰∏∫Á™óÂè£ÁöÑ‰∏≠ÂøÉÊéß‰ª∂Ôºå‰∏≠ÂøÉÊéß‰ª∂ÊòØ‰∏Ä‰∏™Â∏ÉÂ±ÄÔºåÂç≥ view ‰ºöÈöèÁùÄÁ™óÂè£Â§ßÂ∞èÂèòÂåñËá™Âä®ÊîπÂèòÂ∞∫ÂØ∏
    view->setMinimumSize(630, 420); // ÊúÄÂ∞èÂ∞∫ÂØ∏
    view->setAlignment(Qt::AlignLeft | Qt::AlignTop);   // scene Ë∂ÖÂá∫ view ÁöÑËßÜÈáéÊó∂Â±ïÁ§∫Âå∫ÂüüÂØπÈΩêÂà∞ scene ÁöÑÂ∑¶‰∏äÊñπÂå∫Âüü
    view->setRenderHint(QPainter::Antialiasing);
    this->resize(800, 600);         // ÂàùÂßãÂåñÁ™óÂè£Â§ßÂ∞èÔºåÊ≥®ÊÑèÔºöÊï¥‰∏™Á™óÂè£Èô§‰∫Ü‰∏≠ÂøÉÂå∫ÂüüËøòÂåÖÂê´‰∫Ü menubar, toolbar Âíå statusbar Á≠âÂå∫Âüü

    // ÂàùÂßãÂåñÁîªÁ¨î„ÄÅÁîªÂà∑
    pen = new QPen();
    scene->setPen(pen);
    brush = new QBrush(*(new QColor("black")), Qt::NoBrush);
    scene->setBrush(brush);

    init_action_groups();
    init_status_bar();

    // Èöè‰æøÂä†ÁÇπÂàùÂßãÊñáÂ≠ó
    QGraphicsSimpleTextItem* textItem = new QGraphicsSimpleTextItem();
    textItem->setText(QString("‰∏Ä‰∏™Ëø∑‰Ω†ÁîªÂõæÂ∞èÁ®ãÂ∫èüòâ"));
    textItem->setFont(QFont("Times", 12, QFont::Normal));
    textItem->setPos(QPoint(20, 20));
    textItem->setPen(*pen);
    textItem->setBrush(*(new QBrush(*(new QColor("black")), Qt::SolidPattern)));
    scene->addItem(textItem);

}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::resizeEvent(QResizeEvent *event)
{
    Q_UNUSED(event);                                                // Â£∞ÊòéÊú™‰ΩøÁî®ÂèòÈáèÔºåÈÅøÂÖçÁºñËØëÂô®Ë≠¶Âëä
    scene->setSceneRect(0, 0, view->width()-2, view->height()-8);   // Ëã• scene ‰∏é view Â§ßÂ∞èÊé•Ëøë‰ºöËß¶Âèë scrollbarÔºåÁªèÂ∞ùËØïÂáèÂéªËøô‰∏§‰∏™Êï∞ÂàöÂ•ΩËÉΩÊ∂àÂéª
    scene->setRect(QRect(0, 0, view->width(), view->height()));     // Êõ¥Êñ∞ÈúÄË¶ÅÂà∑Êñ∞ÁªòÂõæÁöÑÁü©ÂΩ¢Âå∫Âüü
}

void MainWindow::init_action_groups()
{
    // ÂàÜÁªÑÂèØ‰øùËØÅ‰∏Ä‰∏™ÂàÜÁªÑÂÜÖÂú®Âêå‰∏ÄÊó∂Èó¥Âè™Êúâ‰∏Ä‰∏™ÊåâÈíÆÂèØË¢´Êåâ‰∏ã
    // ÊâÄÊúâÁöÑÁîªÂõæÊ®°ÂºèÂêàÂπ∂‰∏∫‰∏ÄÁªÑÔºå‰∏çËÆæÈªòËÆ§Ê®°Âºè
    QActionGroup *typeGroup = new QActionGroup(this);
    typeGroup->addAction(ui->actionLine);
    typeGroup->addAction(ui->actionRectangle);
    typeGroup->addAction(ui->actionEllipse);
    typeGroup->addAction(ui->actionPolygon);
    typeGroup->addAction(ui->actionFree);
    typeGroup->addAction(ui->actionText);
    typeGroup->addAction(ui->actionEraser);

    // ÊâÄÊúâÁîªÁ¨îÁ∫øÂûãÂêàÂπ∂‰∏∫‰∏ÄÁªÑÔºåÂπ∂ËÆæÂÆûÁ∫ø‰∏∫ÈªòËÆ§Á∫øÂûã
    QActionGroup *lineStyleGroup = new QActionGroup(this);
    lineStyleGroup->addAction(ui->actionSolidLine);
    lineStyleGroup->addAction(ui->actionDashLine);
    lineStyleGroup->addAction(ui->actionDotLine);
    lineStyleGroup->addAction(ui->actionDashDotLine);
    lineStyleGroup->addAction(ui->actionDashDotDotLine);
    lineStyleGroup->addAction(ui->actionNoPen);
    ui->actionSolidLine->setChecked(true);

    // ÊâÄÊúâÂ°´ÂÖÖÁ±ªÂûãÂêàÂπ∂‰∏∫‰∏ÄÁªÑÔºåÂπ∂ËÆæÊó†Â°´ÂÖÖ‰∏∫ÈªòËÆ§Á±ªÂûã
    QActionGroup *brushStyleGroup = new QActionGroup(this);
    brushStyleGroup->addAction(ui->actionNoBrush);
    brushStyleGroup->addAction(ui->actionSolidPattern);
    brushStyleGroup->addAction(ui->actionDense1);
    brushStyleGroup->addAction(ui->actionDense2);
    brushStyleGroup->addAction(ui->actionDense3);
    brushStyleGroup->addAction(ui->actionDense4);
    brushStyleGroup->addAction(ui->actionDense5);
    brushStyleGroup->addAction(ui->actionDense6);
    brushStyleGroup->addAction(ui->actionDense7);
    brushStyleGroup->addAction(ui->actionHorizontal);
    brushStyleGroup->addAction(ui->actionVertical);
    brushStyleGroup->addAction(ui->actionCross);
    brushStyleGroup->addAction(ui->actionForward);
    brushStyleGroup->addAction(ui->actionBackward);
    brushStyleGroup->addAction(ui->actionDiagCross);
    ui->actionNoBrush->setChecked(true);
}

void MainWindow::init_status_bar()
{
    // ‰∏ªÁ™óÂè£‰∏ãÊñπÁä∂ÊÄÅÊ†èÊñáÂ≠óËÆæÁΩÆÂèäÂàùÂßãÂåñ
    status_text_pen_color = new QLabel((QString)"Line Color: " + pen->color().name());
    status_text_pen_color->setMinimumWidth(150);
    ui->statusbar->addPermanentWidget(status_text_pen_color);

    status_text_pen_style = new QLabel((QString)"Style: " +
                                       pen_styles[QString::number(pen->style()).toInt()]);
    status_text_pen_style->setMinimumWidth(170);
    ui->statusbar->addPermanentWidget(status_text_pen_style);

    status_text_pen_width = new QLabel((QString)"Width: " +
                                       QString::number(pen->widthF()) + "px");
    status_text_pen_width->setMinimumWidth(120);
    ui->statusbar->addPermanentWidget(status_text_pen_width);

    status_text_brush_color = new QLabel((QString)"Brush Color: " + brush->color().name());
    status_text_brush_color->setMinimumWidth(150);
    ui->statusbar->addPermanentWidget(status_text_brush_color);

    status_text_brush_style = new QLabel((QString)"Style: " +
                                         brush_styles[QString::number(brush->style()).toInt()]);
    status_text_brush_style->setMinimumWidth(170);
    ui->statusbar->addPermanentWidget(status_text_brush_style);
}

void MainWindow::update_status_text()
{
    // Êõ¥Êñ∞Áä∂ÊÄÅÊ†èÂ≠óÁ¨¶‰∏≤Ôºå‰∏éÂêÑÂ≠óÁ¨¶‰∏≤ÂÖ≥ËÅîÁöÑÁä∂ÊÄÅÊ†è‰ø°ÊÅØ‰ºöËá™Âä®Êõ¥Êñ∞ÂÜÖÂÆπ
    status_text_pen_color->setText((QString)"Line Color: " + pen->color().name());
    status_text_pen_style->setText((QString)"Style: " +
                                    pen_styles[QString::number(pen->style()).toInt()]);
    status_text_pen_width->setText((QString)"Width: " +
                                    QString::number(pen->widthF()) + "px");
    status_text_brush_color->setText((QString)"Brush Color: " + brush->color().name());
    status_text_brush_style->setText((QString)"Style: " +
                                     brush_styles[QString::number(brush->style()).toInt()]);
}

void MainWindow::on_actionAbout_Me_triggered()
{
    QMessageBox::about(this, "About Me", QString("<h2>Ê≠¶Èë´</h2>")+
                                         "<p>Â≠¶Âè∑:&nbsp;&nbsp;11925063<p>" +
                                         "<p>ÈÇÆÁÆ±:&nbsp;&nbsp;wuhsin@zju.edu.cn" +
                                         "<p>GitHub:&nbsp;&nbsp;" +
                                         "<a href=\"https://github.com/XinArkh\">" +
                                         "https://github.com/XinArkh</a>"
                       );
}

void MainWindow::on_actionAbout_QT_triggered()
{
    QMessageBox::aboutQt(this, "About QT");
}

void MainWindow::on_actionLine_triggered()
{
    // ËÆæÁΩÆÂΩìÂâçÁîªÂõæÊ®°Âºè
    scene->setType(QString("Line"));
    // Êõ¥Êñ∞Áä∂ÊÄÅÊ†è
    update_status_text();
}

void MainWindow::on_actionRectangle_triggered()
{
    scene->setType(QString("Rectangle"));
    update_status_text();
}

void MainWindow::on_actionEllipse_triggered()
{
    scene->setType(QString("Ellipse"));
    update_status_text();
}

void MainWindow::on_actionPolygon_triggered()
{
    scene->setType(QString("Polygon"));
    update_status_text();
}

void MainWindow::on_actionFree_triggered()
{
    scene->setType(QString("Free"));
    update_status_text();
}

void MainWindow::on_actionText_triggered()
{
    scene->setType(QString("Text"));
    QString input_text = QInputDialog::getText(this,
                                               "Input Drawing Text",
                                               "Enter Text: ");
    scene->setText(input_text);
    update_status_text();
}

void MainWindow::on_actionEraser_triggered()
{
    scene->setType(QString("Eraser"));
    update_status_text();
}

void MainWindow::on_actionCancel_triggered()
{
    scene->deleteOne();
}

void MainWindow::on_actionClear_triggered()
{
    scene->deleteAll();
}

void MainWindow::on_actionFont_triggered()
{
    bool ok;
    QFont font = QFontDialog::getFont(&ok, scene->getFont(), this);
    if (ok) {
        scene->setFont(font);
    }
}

void MainWindow::on_actionLineWidth_triggered()
{
    bool ok;
    QString width_text = QInputDialog::getText(this,
                                               "Set Pen Width",
                                               "Enter Width (pixel): ",
                                               QLineEdit::Normal,
                                               QString::number(pen->widthF()),
                                               &ok);
    if(ok && !width_text.isEmpty())
    {
        pen->setWidthF(width_text.toDouble());
    }
    update_status_text();
}

void MainWindow::on_actionLineColor_triggered()
{
    QColor line_color = QColorDialog::getColor(pen->color(), this);
    if(line_color.isValid())
    {
        pen->setColor(line_color);
        update_status_text();
    }
}

void MainWindow::on_actionNoPen_triggered()
{
    pen->setStyle(Qt::NoPen);
    update_status_text();
}

void MainWindow::on_actionSolidLine_triggered()
{
    pen->setStyle(Qt::SolidLine);
    update_status_text();
}

void MainWindow::on_actionDashLine_triggered()
{
    pen->setStyle(Qt::DashLine);
    update_status_text();
}

void MainWindow::on_actionDotLine_triggered()
{
    pen->setStyle(Qt::DotLine);
    update_status_text();
}

void MainWindow::on_actionDashDotLine_triggered()
{
    pen->setStyle(Qt::DashDotLine);
    update_status_text();
}

void MainWindow::on_actionDashDotDotLine_triggered()
{
    pen->setStyle(Qt::DashDotDotLine);
    update_status_text();
}

void MainWindow::on_actionNoBrush_triggered()
{
    brush->setStyle(Qt::NoBrush);
    update_status_text();
}

void MainWindow::on_actionSolidPattern_triggered()
{
    brush->setStyle(Qt::SolidPattern);
    update_status_text();
}

void MainWindow::on_actionDense1_triggered()
{
    brush->setStyle(Qt::Dense1Pattern);
    update_status_text();
}

void MainWindow::on_actionDense2_triggered()
{
    brush->setStyle(Qt::Dense2Pattern);
    update_status_text();
}

void MainWindow::on_actionDense3_triggered()
{
    brush->setStyle(Qt::Dense3Pattern);
    update_status_text();
}

void MainWindow::on_actionDense4_triggered()
{
    brush->setStyle(Qt::Dense4Pattern);
    update_status_text();
}

void MainWindow::on_actionDense5_triggered()
{
    brush->setStyle(Qt::Dense5Pattern);
    update_status_text();
}

void MainWindow::on_actionDense6_triggered()
{
    brush->setStyle(Qt::Dense6Pattern);
    update_status_text();
}

void MainWindow::on_actionDense7_triggered()
{
    brush->setStyle(Qt::Dense7Pattern);
    update_status_text();
}

void MainWindow::on_actionHorizontal_triggered()
{
    brush->setStyle(Qt::HorPattern);
    update_status_text();
}

void MainWindow::on_actionVertical_triggered()
{
    brush->setStyle(Qt::VerPattern);
    update_status_text();
}

void MainWindow::on_actionCross_triggered()
{
    brush->setStyle(Qt::CrossPattern);
    update_status_text();
}

void MainWindow::on_actionForward_triggered()
{
    brush->setStyle(Qt::FDiagPattern);
    update_status_text();
}

void MainWindow::on_actionBackward_triggered()
{
    brush->setStyle(Qt::BDiagPattern);
    update_status_text();
}

void MainWindow::on_actionDiagCross_triggered()
{
    brush->setStyle(Qt::DiagCrossPattern);
    update_status_text();
}

void MainWindow::on_actionBrushColor_triggered()
{
    QColor brush_color = QColorDialog::getColor(brush->color(), this);
    if(brush_color.isValid())
    {
        brush->setColor(brush_color);
        update_status_text();
    }
}

void MainWindow::on_actionSave_As_triggered()
{
    QString file_path = QFileDialog::getSaveFileName(this,
                                                     "Save As",
                                                     "untitled.png",
                                                     "Images (*.png *.bmp *.jpg)");
    QScreen *screen = QGuiApplication::primaryScreen();
    screen->grabWindow(view->winId()).save(file_path);
}
